clash-prelude/src/Clash/Explicit/BlockRam.hs

```diff
@@ -1064,3 +1071,258 @@ readNew clk rst en ram rdAddr wrM = mux wasSame wasWritten $ ram rdAddr wrM
         (wasSame,wasWritten) =
           unbundle (register clk rst en (False, undefined)
                              (readNewT <$> rdAddr <*> wrM))
+
+-- | Produces vendor-agnostic HDL that will be inferred as a true, dual port
+-- block ram. It has write-first (or "new data") behavior, no clock enables,
+-- no byte enables, and is symmetric in its data ports.
+--
+-- * __NB__: Read value is delayed by 1 cycle
+-- * __NB__: Initial output value is /undefined/
+-- * __NB__: Port A and port B cannot make overlapping read/write requests on
+--           the same address.
+--
+trueDualPortBlockRam ::
+  forall nAddrs domA domB a .
+  ( HasCallStack
+  , KnownNat nAddrs
+  , KnownDomain domA
+  , KnownDomain domB
+  , NFDataX a
+  , BitPack a
+  )
+  => Clock domA
+  -- ^ Clock for port A
+  -> Signal domA (Maybe a)
+  -- ^ Write data port A
+  -> Signal domA (Index nAddrs)
+  -- ^ Address to read from or write to on port A
+
+  -> Clock domB
+  -- ^ Clock for port B
+  -> Signal domB (Maybe a)
```

Comment:
I would rather have a new type for this. I assume if you pass in a `Just` here then it means you want to write and if you pass `Nothing` then a value is read. Why not just make a special `Either` here. `data DualPortRamOp n a = DualPortWrite (Index n) a | DualPortRead (Index n)` or something.